rpc,rest,zmq: faster getblock, NotifyBlock and rest_block by reading raw block

andrewtoth opened the pull request:
For the `getblock` endpoint with `verbosity=0`, the  `rest_block` REST endpoint for `bin` and `hex`, and zmq `NotifyBlock` we don't have to deserialize the block since we're just sending the raw data. This PR uses `ReadRawBlockFromDisk` instead of `ReadBlockFromDisk` to serve these requests, and only deserializes for `verbosity > 0` and `json` REST requests. See benchmarks in https://github.com/bitcoin/bitcoin/pull/26684.

Benchmarked using ApacheBench. Requesting block 750,000 in binary 10k times on a single core (set `-rest=1` in config):
`ab -n 10000 -c 1 "http://127.0.0.1:8332/rest/block/0000000000000000000592a974b1b9f087cb77628bb4a097d5c2c11b3476a58e.bin"`

On master, mean time 15ms.
On this branch, mean time 1ms.

For RPC
```
echo '{"jsonrpc": "1.0", "id": "curltest", "method": "getblock", "params": ["0000000000000000000592a974b1b9f087cb77628bb4a097d5c2c11b3476a58e", 0]}' > /tmp/data.json
ab -p /tmp/data.json -n 1000 -c 1 -A user:password "http://127.0.0.1:8332/"
```
On master, mean time 32ms
On this branch, mean time 13ms

andrewtoth commented:
For some more context, here is the PR that introduced `ReadRawBlockFromDisk` https://github.com/bitcoin/bitcoin/pull/13151.

andrewtoth commented:
@luke-jr I wasn't aware of #21319 until now, but I can't comment there because it's locked. If you want to reopen that I can review and remove the change to RPC in this PR so it will only be for REST.

TheCharlatan commented on src/rest.cpp:
```diff
@@ -37,6 +37,7 @@
 using node::GetTransaction;
```
I'm guessing the commit message should say "rest:..."?

sedited reviewed (commented):
Concept ACK

Code looks good to me, but still want to run some tests.

RandyMcMillan commented:
Concept ACK

andrewtoth commented:
@TheCharlatan @RandyMcMillan rebased.

TheCharlatan commented on src/zmq/zmqnotificationinterface.h:
```diff
@@ -5,6 +5,7 @@
 #ifndef BITCOIN_ZMQ_ZMQNOTIFICATIONINTERFACE_H
 #define BITCOIN_ZMQ_ZMQNOTIFICATIONINTERFACE_H
 
+#include <flatfile.h>
```
In commit 6229158f77a7b1803f2767a089dd94dedf62b533
Nit: You could just forward-declare `struct FlatFilePos` without having to include flatfile (same for `zmqpublishnotifier.h`).

sedited reviewed (commented):
ACK 2ccbef905fb6faa2480b49103680529dc7cfc482

TheCharlatan commented on src/zmq/zmqnotificationinterface.h:
```diff
@@ -17,14 +17,16 @@ class CBlock;
     std::list<const CZMQAbstractNotifier*> GetActiveNotifiers() const;
 
-    static std::unique_ptr<CZMQNotificationInterface> Create(std::function<bool(CBlock&, const CBlockIndex&)> get_block_by_index);
+    static std::unique_ptr<CZMQNotificationInterface> Create(std::function<bool(CBlock&, const CBlockIndex&)> get_block_by_index, std::function<bool(std::vector<uint8_t>&, const FlatFilePos&)> get_raw_block);
```
Nit: `#include <vector>`

sedited reviewed (approved):
Re-ACK a13264fb861139a38b70634d0885101ebaa24ed6

andrewtoth commented:
> Does testing the different branches involve something more than checking out the different branches (master and read-raw-block) on my local computer?

@kashifs after checkout out each branch, you must again compile from source, and shut down the running bitcoind and then run the newly compiled binary at ./src/bitcoind.

kashifs commented:
@andrewtoth, okay thanks. I did as you suggested and was able to get similar speed improvements when testing the `rest` command. 10.260ms on branch `master` and 1.014ms on branch `bitcoin-read-raw-block`. I only see improvements when running the `rest` testing command more than once though. Is this the expected behavior? I hope it's not due to some caching or other improvements unrelated to your code updates. Also, I was not able to appreciate any improvements when running the `RPC` command. Could there be something else that I'm missing? Thanks again for your time.

pablomartin4btc reviewed (commented):
Concept ACK

I'd like to run the bench using #26684 which I have pending of reviewing.

pablomartin4btc commented on src/rpc/blockchain.cpp:
```diff
@@ -736,6 +758,10 @@ static RPCHelpMan getblock()
         }
     }
 
+    if (verbosity <= 0 && !RPCSerializationWithoutWitness()) {
```
Thought was due to #28438 & #28448... but just found #28890, nice! This PR required rebase anyways.

andrewtoth replied:
Rebased

andrewtoth commented on src/rpc/blockchain.cpp:
```diff
@@ -594,6 +595,28 @@ static CBlock GetBlockChecked(BlockManager& blockman, const CBlockIndex& blockin
+        pos = blockindex.GetBlockPos();
+    }
+
+    if (!blockman.ReadRawBlockFromDisk(data, pos)) {
```
Fixed. Apparently `IsBlockPruned` is not sufficient to know whether we can read the block or not. Getting the file position of a block that we don't have returns us a null `FlatFilePos`, and `ReadBlockFromDisk` fails on opening a file at null position. `ReadRawBlockFromDisk` instead first tries to decrement `nPos` by 8 before trying to open the file, which results in UB due to `nPos` being unsigned and 0 for a null position.

TheCharlatan commented on src/node/blockstorage.cpp:
```diff
@@ -1067,6 +1067,11 @@ bool BlockManager::ReadBlockFromDisk(CBlock& block, const CBlockIndex& index) co
+    // Return early to prevent undefined behavior of unsigned int underflow
+    if (hpos.nPos < 8) {
+        return error("%s: OpenBlockFile failed for %s", __func__, pos.ToString());
+    }
```
This check seems like the right thing to do, but this could be handled by the caller, by checking `if (!(pindex->nStatus & BLOCK_HAVE_DATA))`, like done in `net_processing.cpp:2292`. Maybe still add this in a separate commit to clearly explain what is going and why being defensive here is an improvement?

andrewtoth replied:
Done, added as a separate commit at the start of this PR.
Yes, it could be handled by the caller, but it's really handled by `OpenBlockFile` which checks for a null `pos`. It's just that before that check it blindly decrements the nPos by 8. Now we check that and return the same error that would be returned by `OpenBlockFile` if `pos` is null.

furszy commented on src/rpc/blockchain.cpp:
```diff
@@ -594,6 +595,28 @@ static CBlock GetBlockChecked(BlockManager& blockman, const CBlockIndex& blockin
 
+static std::vector<uint8_t> GetRawBlockChecked(BlockManager& blockman, const CBlockIndex& blockindex)
+{
+    std::vector<uint8_t> data{};
+    FlatFilePos pos{};
+    {
+        LOCK(cs_main);
+        if (blockman.IsBlockPruned(blockindex)) {
+            throw JSONRPCError(RPC_MISC_ERROR, "Block not available (pruned data)");
+        }
+        pos = blockindex.GetBlockPos();
+    }
```
As we only care about the data existence here, what about doing a more general check:
```c++
if (!(blockindex.nStatus & BLOCK_HAVE_DATA)) {
    throw JSONRPCError(RPC_MISC_ERROR, "Block not available");
}
```

(Unless you want to keep the same error message as before)

andrewtoth replied:
I think this check is specifically done to see if the block is pruned, hence the error message `(pruned data)`. This is taken verbatim from the above function, `GetBlockChecked`, which all it does inside the locked block is check if it is pruned or not. Here we also extract the `FlatFilePos`. However, if `blockindex.nStatus & BLOCK_HAVE_DATA` is `false` then the call to `ReadRawBlockFromDisk` will fail and throw the message `Block not available`. The first commit addresses the undefined behavior part of calling `ReadRawBlockFromDisk` with a null `FlatFilePos`, which is safe to do now.

furszy commented on src/rpc/blockchain.cpp:
```diff
@@ -734,15 +757,17 @@ static RPCHelpMan getblock()
         }
     }
 
-    const CBlock block{GetBlockChecked(chainman.m_blockman, *pblockindex)};
+    const std::vector<uint8_t> block_data{GetRawBlockChecked(chainman.m_blockman, *pblockindex)};
 
     if (verbosity <= 0) {
-        DataStream ssBlock;
-        ssBlock << TX_WITH_WITNESS(block);
-        std::string strHex = HexStr(ssBlock);
+        std::string strHex{HexStr(block_data)};
         return strHex;
```
In 3f0f43d605b44:

could inline it:
```c++
if (verbosity <= 0) {
    return HexStr(block_data);
}
```

andrewtoth replied:
Done

furszy commented on src/rpc/blockchain.cpp:
```diff
@@ -734,15 +757,17 @@ static RPCHelpMan getblock()
     }
 
-    const CBlock block{GetBlockChecked(chainman.m_blockman, *pblockindex)};
+    const std::vector<uint8_t> block_data{GetRawBlockChecked(chainman.m_blockman, *pblockindex)};
```
In 3f0f43d60:

A topic about this change:

`GetBlockChecked()` internally checks the block proof of work and the signet signature after reading the data from disk. The new `GetRawBlockChecked()` does not perform this checks.

It isn't an issue for me because the block wouldn't have been stored if the PoW was invalid and I don't think that `getblock` is the appropriate place to check for corruption or tampered data.

andrewtoth replied:
Done, added a commit description addressing this.

furszy commented on src/rest.cpp:
```diff
@@ -307,32 +308,33 @@ static bool rest_block(const std::any& context,
     }
 
-    if (!chainman.m_blockman.ReadBlockFromDisk(block, *pblockindex)) {
+    std::vector<uint8_t> block_data{};
+    if (!chainman.m_blockman.ReadRawBlockFromDisk(block_data, pos)) {
```
In 3d7ae60f21:

Same discussion as with the previous commit. `ReadBlockFromDisk()` checks PoW and the signet signature after reading the block from disk while `ReadRawBlockFromDisk()` doesn't.

I agree on not performing these checks but would be nice to describe the changes in the commit description.

andrewtoth replied:
Done, added a commit description addressing this.

andrewtoth commented:
@furszy @TheCharlatan updated with both your suggestions, thanks!

furszy reviewed (commented):
Code ACK 4fc91b38

sedited reviewed (approved):
Re-ACK 4fc91b38ae8c715e0ebf9abcb6da17044fe418b3

achow101 commented:
ACK c0a2294d3cd521dbb8bd87a63273c42ae1b889b7

achow101 commented on src/node/blockstorage.cpp:
```diff
@@ -1067,6 +1067,11 @@ bool BlockManager::ReadBlockFromDisk(CBlock& block, const CBlockIndex& index) co
+    // If nPos is less than 8 the pos is null and we don't have the block data
+    // Return early to prevent undefined behavior of unsigned int underflow
+    if (hpos.nPos < 8) {
+        return error("%s: OpenBlockFile failed for %s", __func__, pos.ToString());
```
In 3f086a9353191a5df03aa595bdbb42c2195a23d2 "blockstorage: check nPos in ReadRawBlockFromDisk before seeking back"

Silent merge conflict here:

```
../../../src/node/blockstorage.cpp: In member function ‘bool node::BlockManager::ReadRawBlockFromDisk(std::vector<unsigned char>&, const FlatFilePos&) const’:
../../../src/node/blockstorage.cpp:1094:16: error: ‘error’ was not declared in this scope; did you mean ‘perror’?
 1094 |         return error("%s: OpenBlockFile failed for %s", __func__, pos.ToString());
      |                ^~~~~
      |                perror
```

`error()` was removed in #29236

andrewtoth replied:
@achow101 resolved, thanks.

andrewtoth committed:
blockstorage: check nPos in ReadRawBlockFromDisk before seeking back

ReadRawBlockFromDisk assumes a non-null pos that has an nPos >= 8.
This simple check makes the function safer to call in the future,
so callers don't need to worry about causing UB if the pos is null.

andrewtoth committed:
zmq: read raw block with ReadRawBlockFromDisk

andrewtoth committed:
test: check more details on zmq raw block response

andrewtoth committed:
rpc: read raw block in getblock and deserialize for verbosity > 0

Note that for speed this commit also removes the proof of work and
signet signature checks before returning the block in getblock.
It is assumed if a block is stored it will be valid.

andrewtoth committed:
rest: read raw block in rest_block and deserialize for json

Note that for speed this commit also removes the proof of work and
signet signature checks before returning the block in getblock.
It is assumed if a block is stored it will be valid.

achow101 commented:
re-ACK e710cefd5701cd33d1e55034b3e37cea78582733

Verified rebase diff.
